solvers/bcd_celer.py

```diff
@@ -0,0 +1,363 @@
+        else:
+            anderson /= np.sum(anderson)
+
+            out[:] = 0
+            for k in range(K - 1):
+                out += anderson[k] * last_K_R[k, :]
+            out /= alpha
+            # out now holds the extrapolated dual point
+
+
+def bcd_epoch(C, norms_X_block, X, R, alpha, W, inv_lc):
+    n_tasks = R.shape[1]
+    W_j_new = np.zeros((1, n_tasks))
+    dgemm = _get_dgemm()
+    alpha_lc = alpha * inv_lc
+
+    for j in C:
+        if norms_X_block[j] == 0.:
+            continue
+        idx = slice(j, j+1)
+        W_j = W[idx, :]
+        X_j = X[:, idx]
+
+        # W_j_old = W[j].copy()
+
+        # W_j_new = X_j.T @ R * inv_lc[j]
+        dgemm(alpha=inv_lc[j], beta=0.0, a=R.T, b=X_j, c=W_j_new.T,
+              overwrite_c=True)
+
+        if W_j[0, 0] != 0:
+            # R += np.dot(X_j, W_j)
+            dgemm(alpha=1.0, beta=1.0, a=W_j.T, b=X_j.T, c=R.T,
+                  overwrite_c=True)
+            W_j_new += W_j
+
+        # block_norm = np.sqrt(sum_squared(W_j_new))
+        block_norm = norm(W_j_new)
+
+        if block_norm <= alpha_lc[j]:
+            W_j.fill(0.0)
+        else:
+            shrink = 1.0 - alpha_lc[j] / block_norm
+            W_j_new *= shrink
+            # R -= np.dot(X_j, W_j_new)
+            dgemm(alpha=-1.0, beta=1.0, a=W_j_new.T, b=X_j.T, c=R.T,
+                  overwrite_c=True)
+            W_j[:] = W_j_new
+
+        # R += np.outer(X[:, j], W_j_old - W[j])
+        # R[:] = Y - X @ W
+        # np.testing.assert_allclose(Y - X @ W, R)
+
+
+def celer_dual_mtl(X, Y, alpha, n_iter, max_epochs=10_000, gap_freq=10,
+                   tol=1e-12, p0=10, verbose=0, prune=True, accel=True):
+    n_samples, n_features = X.shape
+    n_tasks = Y.shape[1]
+    n_obs = n_samples * n_tasks
+
+    W = np.zeros((n_features, n_tasks))
+    R = Y.copy()
+    verbose_in = max(verbose - 1, 0)
+    n_screened = 0
+    norm_Y2 = norm(Y, 'fro') ** 2
+
+    tol *= norm_Y2
+    if p0 > n_features:
+        p0 = n_features
+
+    prios = np.empty(n_features, dtype=X.dtype)
+    screened = np.zeros(n_features, dtype=np.int32)
+    notin_WS = np.zeros(n_features, dtype=np.int32)
+
+    # acceleration variables
+    K = 6
+    last_K_R = np.empty((K, n_samples, n_tasks), dtype=X.dtype)
+    U = np.empty((K - 1, n_obs), dtype=X.dtype)
+    UtU = np.empty((K - 1, K - 1), dtype=X.dtype)
+
+    norms_X_block = norm(X, axis=0)
+    inv_lc = 1 / norms_X_block ** 2
+
+    gaps = np.zeros(n_iter, dtype=X.dtype)
+
+    Theta = np.zeros((n_samples, n_tasks), dtype=X.dtype)
+    Theta_in = np.zeros((n_samples, n_tasks), dtype=X.dtype)
+    Thetacc = np.zeros((n_samples, n_tasks), dtype=X.dtype)
+
+    # d_obj_from_inner = 0.
+    all_features = np.arange(n_features)
+    C = all_features.copy()
+    ws_size = p0
+
+    for t in range(n_iter):
+        Theta[:] = R / alpha
+
+        norm_XT_Theta = norm(X.T @ Theta, axis=1)
+        scal = max(norm_XT_Theta)
+        # The piece of code below is slow.
```

Comment:
in the end did you find out why ?

Reply:
because X[:, j] @ Theta is slow ? and how come doing it for the full X is faster, I am puzzled

Reply:
Because dual_scaling_mtl is (was?) slow.

Reply:
yes, but *why* was it slow ?

Reply:
I think it is still slow, but not the most consuming task when we profiled. I'll send you a report of profiled code when I'm done.